packages/techdocs-common/src/stages/publish/azureStorage.ts

```diff
@@ -0,0 +1,223 @@
+    this.containerName = containerName;
+    this.logger = logger;
+  }
+
+  /**
+   * Upload all the files from the generated `directory` to the Azure Storage container.
+   * Directory structure used in the container is - entityNamespace/entityKind/entityName/index.html
+   */
+  async publish({ entity, directory }: PublishRequest): Promise<void> {
+    try {
+      // Note: Azure Storage manages creation of parent directories if they do not exist.
+      // So collecting path of only the files is good enough.
+      const allFilesToUpload = await getFileTreeRecursively(directory);
+
+      const uploadPromises: Array<Promise<BlobUploadCommonResponse>> = [];
+      allFilesToUpload.forEach(filePath => {
+        // Remove the absolute path prefix of the source directory
+        // Path of all files to upload, relative to the root of the source directory
+        // e.g. ['index.html', 'sub-page/index.html', 'assets/images/favicon.png']
+        const relativeFilePath = filePath.replace(`${directory}/`, '');
+        const entityRootDir = `${entity.metadata.namespace}/${entity.kind}/${entity.metadata.name}`;
+        const destination = path.normalize(
+          `${entityRootDir}/${relativeFilePath}`,
+        ); // Azure Storage Container file relative path
+        // TODO: Upload in chunks of ~10 files instead of all files at once.
+        uploadPromises.push(
+          this.storageClient
+            .getContainerClient(this.containerName)
+            .getBlockBlobClient(destination)
+            .uploadFile(filePath),
+        );
+      });
+
+      await Promise.all(uploadPromises).then(() => {
+        this.logger.info(
+          `Successfully uploaded all the generated files for Entity ${entity.metadata.name}. Total number of files: ${allFilesToUpload.length}`,
+        );
+      });
+      return;
+    } catch (e) {
+      const errorMessage = `Unable to upload file(s) to Azure Storage. Error ${e.message}`;
+      this.logger.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+  }
+
+  download(containerName: string, path: string): Promise<string> {
+    return new Promise((resolve, reject) => {
+      const fileStreamChunks: Array<any> = [];
+      this.storageClient
+        .getContainerClient(containerName)
+        .getBlockBlobClient(path)
+        .download()
+        .then(res => {
+          const body = res.readableStreamBody;
+          if (!body) {
+            reject(new Error(`Unable to parse the response data`));
+            return;
+          }
+          body
+            .on('error', e => {
+              this.logger.error(e.message);
+              reject(e.message);
+            })
+            .on('data', chunk => {
+              fileStreamChunks.push(chunk);
+            })
+            .on('end', () => {
+              resolve(Buffer.concat(fileStreamChunks).toString());
+            });
+        });
+    });
+  }
+
+  async fetchTechDocsMetadata(entityName: EntityName): Promise<string> {
```

Comment:
Just a heads up that https://github.com/backstage/backstage/pull/3968 should be merged very soon, which introduces the Type for the metadata. So, instead of `Promise<string>`, we'll do `Promise<TechDocsMetadata>` here. Look at the PR for hints on how AWS/GCS publishers do this.

Reply:
Changed